nimare/meta/mkda.py

```diff
@@ -192,63 +188,15 @@ def _compute_null_analytic(self, ma_maps):
-        self.null_distributions_["histogram_bins"] = np.linspace(0, bins_max, num=N_BINS)
-
-        ma_hists = np.zeros(
-            (ma_values.shape[0], self.null_distributions_["histogram_bins"].shape[0])
-        )
-        for i_exp in range(ma_values.shape[0]):
-            # Remember that histogram uses bin edges (not centers), so it
-            # returns a 1xhist_bins-1 array
-            n_zeros = len(np.where(ma_values[i_exp, :] == 0)[0])
-            reduced_ma_values = ma_values[i_exp, ma_values[i_exp, :] > 0]
-            ma_hists[i_exp, 0] = n_zeros
-            ma_hists[i_exp, 1:] = np.histogram(
-                a=reduced_ma_values, bins=self.null_distributions_["histogram_bins"], density=False
-            )[0]
-
-        # Inverse of step size in histBins (0.0001) = 10000
-        step = 1 / np.mean(np.diff(self.null_distributions_["histogram_bins"]))
-
-        # Null distribution to convert ALE to p-values.
-        stat_hist = ma_hists[0, :]
-        for i_exp in range(1, ma_hists.shape[0]):
-            temp_hist = np.copy(stat_hist)
-            ma_hist = np.copy(ma_hists[i_exp, :])
-
-            # Find histogram bins with nonzero values for each histogram.
-            ale_idx = np.where(temp_hist > 0)[0]
-            exp_idx = np.where(ma_hist > 0)[0]
-
-            # Normalize histograms.
-            temp_hist /= np.sum(temp_hist)
-            ma_hist /= np.sum(ma_hist)
-
-            # Perform weighted convolution of histograms.
-            stat_hist = np.zeros(self.null_distributions_["histogram_bins"].shape[0])
-            for j_idx in exp_idx:
-                # Compute probabilities of observing each ALE value in histBins
-                # by randomly combining maps represented by maHist and aleHist.
-                # Add observed probabilities to corresponding bins in ALE
-                # histogram.
-                probabilities = ma_hist[j_idx] * temp_hist[ale_idx]
-                ale_scores = 1 - (1 - self.null_distributions_["histogram_bins"][j_idx]) * (
-                    1 - self.null_distributions_["histogram_bins"][ale_idx]
-                )
-                score_idx = np.floor(ale_scores * step).astype(int)
-                np.add.at(stat_hist, score_idx, probabilities)
-
-        # Convert aleHist into null distribution. The value in each bin
-        # represents the probability of finding an ALE value (stored in
-        # histBins) of that value or lower.
-        null_distribution = stat_hist / np.sum(stat_hist)
-        null_distribution = np.cumsum(null_distribution[::-1])[::-1]
+        # MKDA maps are binary, so we only have k + 1 bins in the final
+        # histogram, where k is the number of studies. We can analytically
+        # compute the null distribution by convolution.
+        prop_active = ma_values.mean(1)
+        ss_hist = 1.
+        for exp_prop in prop_active:
+            ss_hist = np.convolve(ss_hist, [1 - exp_prop, exp_prop])
```

Comment:
Whoa... that's seriously awesome.

Reply:
It gives different results from the old version but there is at least one major bug in the old version ([here](https://github.com/neurostuff/NiMARE/blob/44c493a0b71a85b8aeae297a7f566bad532035c7/nimare/meta/mkda.py#L241-L243)), so I can't really consider backwards incompatibility a problem.

Reply:
It converges almost perfectly with the Monte Carlo version (see [this new test](https://github.com/neurostuff/NiMARE/blob/69f1aada8fc24d9baea3e50ea1330bdb31305d56/nimare/tests/test_meta_mkda.py#L119), which has a pretty low tolerance for difference), so I'm pretty confident that they're both right—or at least, any bug would have to be in the common code, likely in `compute_summarystat` (which is possible—I did modify it). :)